test/unit/discovered_extensions_test.rb

```diff
@@ -95,12 +95,91 @@ class FindDiscoveryRulesTest < ActiveSupport::TestCase
     assert_equal host.name, "macaabbccddeeff"
   end
 
-  test "hostname is rendered after auto provisioning" do
-    facts = @facts.merge({"somefact" => "abc"})
-    host = Host::Discovered.import_host_and_facts(facts).first
-    r1 = FactoryGirl.create(:discovery_rule, :priority => 1, :search => "facts.somefact = abc", :hostname => 'x<%= 1+1 %>')
-    perform_auto_provision host, r1
-    assert_equal host.name, "x2"
+  def setup_normal_renderer
+    Setting[:safemode_render] = false
+  end
+
+  def setup_safemode_renderer
+    Setting[:safemode_render] = true
+  end
+
+  [:normal_renderer, :safemode_renderer].each do |renderer_name|
+    test "#{renderer_name} is properly configured" do
+      send "setup_#{renderer_name}"
+      if renderer_name == :normal_renderer
+        assert Setting[:safemode_render] == false
+      else
+        assert Setting[:safemode_render] == true
+      end
+    end
+
+    test "hostname falls back to original name on empty response via #{renderer_name}" do
+      facts = @facts.merge({"somefact" => "abc"})
+      host = Host::Discovered.import_host_and_facts(facts).first
+      r1 = FactoryGirl.create(:discovery_rule,
+                              :priority => 1,
+                              :search => "facts.somefact = abc",
+                              :hostname => '<%= "" %>')
+      perform_auto_provision host, r1
+      assert_equal "macaabbccddeeff", host.name
+    end
+
+    test "hostname is rendered after auto provisioning using #{renderer_name}" do
+      facts = @facts.merge({"somefact" => "abc"})
+      host = Host::Discovered.import_host_and_facts(facts).first
+      r1 = FactoryGirl.create(:discovery_rule,
```

Comment:
@lzap why are you adding the priority and search if they already exist in the factory?
maybe change the fact in the factory or the fact here and then you can just ad `hostname`?
you could also move the `@facts.merge` to the setup